noop/shared/src/main/scala/org/typelevel/log4cats/noop/NoOpLogger.scala

```diff
@@ -21,39 +21,51 @@ import org.typelevel.log4cats.SelfAwareStructuredLogger
 
 object NoOpLogger {
   def apply[F[_]: Applicative]: SelfAwareStructuredLogger[F] = impl[F]
-  def impl[F[_]: Applicative]: SelfAwareStructuredLogger[F] = new SelfAwareStructuredLogger[F] {
 
-    val no: F[Boolean] = Applicative[F].pure(false)
-    val unit: F[Unit] = Applicative[F].pure(())
-    @inline override def isTraceEnabled: F[Boolean] = no
-    @inline override def isDebugEnabled: F[Boolean] = no
-    @inline override def isInfoEnabled: F[Boolean] = no
-    @inline override def isWarnEnabled: F[Boolean] = no
-    @inline override def isErrorEnabled: F[Boolean] = no
-    @inline override def trace(t: Throwable)(msg: => String): F[Unit] = unit
-    @inline override def trace(msg: => String): F[Unit] = unit
-    @inline override def trace(ctx: Map[String, String])(msg: => String): F[Unit] = unit
-    @inline override def debug(t: Throwable)(msg: => String): F[Unit] = unit
-    @inline override def debug(msg: => String): F[Unit] = unit
-    @inline override def debug(ctx: Map[String, String])(msg: => String): F[Unit] = unit
-    @inline override def info(t: Throwable)(msg: => String): F[Unit] = unit
-    @inline override def info(msg: => String): F[Unit] = unit
-    @inline override def info(ctx: Map[String, String])(msg: => String): F[Unit] = unit
-    @inline override def warn(t: Throwable)(msg: => String): F[Unit] = unit
-    @inline override def warn(msg: => String): F[Unit] = unit
-    @inline override def warn(ctx: Map[String, String])(msg: => String): F[Unit] = unit
-    @inline override def error(t: Throwable)(msg: => String): F[Unit] = unit
-    @inline override def error(msg: => String): F[Unit] = unit
-    @inline override def error(ctx: Map[String, String])(msg: => String): F[Unit] = unit
-    @inline override def trace(ctx: Map[String, String], t: Throwable)(msg: => String): F[Unit] =
-      unit
-    @inline override def debug(ctx: Map[String, String], t: Throwable)(msg: => String): F[Unit] =
-      unit
-    @inline override def info(ctx: Map[String, String], t: Throwable)(msg: => String): F[Unit] =
-      unit
-    @inline override def warn(ctx: Map[String, String], t: Throwable)(msg: => String): F[Unit] =
-      unit
-    @inline override def error(ctx: Map[String, String], t: Throwable)(msg: => String): F[Unit] =
-      unit
-  }
+  def strictEvalArgs[F[_]: Applicative]: SelfAwareStructuredLogger[F] =
+    impl_[F](evaluateArgs = true)
+
+  def impl[F[_]: Applicative] = impl_[F](evaluateArgs = false)
+
+  private[noop] def impl_[F[_]: Applicative](evaluateArgs: Boolean): SelfAwareStructuredLogger[F] =
+    new SelfAwareStructuredLogger[F] {
+
+      val no: F[Boolean] = Applicative[F].pure(false)
+      def void(arg: => Any): F[Unit] = Applicative[F].pure {
```

Comment:
What if we did the strictness check outside, and returned either an unconditionally strict or unconditionally lazy implementation?  Then we could avoid the conditional on each call (probably insignificant).  More importantly, we'd keep the F.unit optimization in the lazy case.

Reply:
Good point. Will do that